### ISSUE TITLE
gosec(G602): false positives

### ISSUE THREAD

**Issue by timbray** [golangci/golangci-lint#6334]
### Welcome

- [x] Yes, I'm using a binary release within 2 latest [releases](https://github.com/golangci/golangci-lint/releases). Only such installations are supported.
- [x] Yes, I've searched similar [issues on GitHub](https://github.com/golangci/golangci-lint/issues) and didn't find any.
- [x] Yes, I've read the `typecheck` section of the [FAQ](https://golangci-lint.run/docs/welcome/faq/#why-do-you-have-typecheck-errors).
- [x] Yes, I've tried with the standalone [linter](https://golangci-lint.run/docs/linters/) if available (e.g., gocritic, go vet, etc.).
- [x] I agree to follow this project's [Code of Conduct](https://github.com/golangci/golangci-lint?tab=coc-ov-file)

### How did you install golangci-lint?

Brew

### Description of the problem

Big `switch` statement is reporting G602 on all the cases except for the last.  I'm pretty sure the slice index isn't out of range, I guess for some reason golangci-lint doesn't see the make([]byte) call?

```go
func (fj *flattenJSON) readTextWithEscapes(from int) ([]byte, int, error) {
	// pointing at \
	unescaped := make([]byte, 1)
	var err error
	from++
	if from == len(fj.event) {
		return nil, 0, fj.error("premature end of event")
	}
	switch fj.event[from] {
	case '"':
		unescaped[0] = '"'
	case '\\':
		unescaped[0] = '\\'
	case '/':
		unescaped[0] = '/'
	case 'b':
		unescaped[0] = 8
	case 'f':
		unescaped[0] = 0xc
	case 'n':
		unescaped[0] = '\n'
	case 'r':
		unescaped[0] = '\r'
	case 't':
		unescaped[0] = '\t'
	case 'u':
		unescaped, from, err = fj.readHexUTF16(from)
		if err != nil {
			return nil, 0, err
		}
	default:
		return nil, 0, fj.error("malformed \\-escape in text")
	}

	return unescaped, from, nil
}
```

### Version of golangci-lint

<details>

```console
$ golangci-lint --version
golangci-lint has version 2.8.0 built with go1.25.5 from e2e40021 on 2026-01-07T21:29:47Z
```

</details>

### Configuration

<details>

```yml
version: "2"
linters:
  enable:
    - bodyclose
    - errname
    - errorlint
    - gochecknoinits
    - goprintffuncname
    - gosec
    - misspell
    - nilerr
    - nilnil
    - predeclared
    - staticcheck
    - thelper
    - tparallel
    - unparam
    - whitespace
  exclusions:
    generated: lax
    presets:
      - comments
      - common-false-positives
      - legacy
      - std-error-handling
    rules:
      - path: _test\.go
        text: Use of weak random number generator
      - path: _test\.go
        text: ST1018
    paths:
      - third_party$
      - builtin$
      - examples$
formatters:
  enable:
    - gofmt
    - goimports
  exclusions:
    generated: lax
    paths:
      - third_party$
      - builtin$
      - examples$
```

</details>

### Go environment

<details>

```console
$ go version && go env
go version go1.25.6 darwin/arm64
AR='ar'
CC='cc'
CGO_CFLAGS='-O2 -g'
CGO_CPPFLAGS=''
CGO_CXXFLAGS='-O2 -g'
CGO_ENABLED='1'
CGO_FFLAGS='-O2 -g'
CGO_LDFLAGS='-O2 -g'
CXX='c++'
GCCGO='gccgo'
GO111MODULE=''
GOARCH='arm64'
GOARM64='v8.0'
GOAUTH='netrc'
GOBIN=''
GOCACHE='/Users/timothybray/Library/Caches/go-build'
GOCACHEPROG=''
GODEBUG=''
GOENV='/Users/timothybray/Library/Application Support/go/env'
GOEXE=''
GOEXPERIMENT=''
GOFIPS140='off'
GOFLAGS=''
GOGCCFLAGS='-fPIC -arch arm64 -pthread -fno-caret-diagnostics -Qunused-arguments -fmessage-length=0 -ffile-prefix-map=/var/folders/7x/gb0mpjt17_x97kz0k1gn449c0000gn/T/go-build2149191522=/tmp/go-build -gno-record-gcc-switches -fno-common'
GOHOSTARCH='arm64'
GOHOSTOS='darwin'
GOINSECURE=''
GOMOD='/Users/timothybray/dev/quamina/go.mod'
GOMODCACHE='/Users/timothybray/go/pkg/mod'
GONOPROXY=''
GONOSUMDB=''
GOOS='darwin'
GOPATH='/Users/timothybray/go'
GOPRIVATE=''
GOPROXY='https://proxy.golang.org,direct'
GOROOT='/opt/homebrew/Cellar/go/1.25.6/libexec'
GOSUMDB='sum.golang.org'
GOTELEMETRY='local'
GOTELEMETRYDIR='/Users/timothybray/Library/Application Support/go/telemetry'
GOTMPDIR=''
GOTOOLCHAIN='auto'
GOTOOLDIR='/opt/homebrew/Cellar/go/1.25.6/libexec/pkg/tool/darwin_arm64'
GOVCS=''
GOVERSION='go1.25.6'
GOWORK=''
PKG_CONFIG='pkg-config'
```

</details>

### Verbose output of running

<details>

```console
$ golangci-lint cache clean
$ golangci-lint run -v
INFO golangci-lint has version 2.8.0 built with go1.25.5 from e2e40021 on 2026-01-07T21:29:47Z 
INFO [config_reader] Config search paths: [./ /Users/timothybray/dev/quamina /Users/timothybray/dev /Users/timothybray /Users /] 
INFO [config_reader] Used config file .golangci.yaml 
INFO [config_reader] Module name "quamina.net/go/quamina" 
INFO maxprocs: Leaving GOMAXPROCS=12: CPU quota undefined 
INFO [goenv] Read go env for 7.459791ms: map[string]string{"GOCACHE":"/Users/timothybray/Library/Caches/go-build", "GOROOT":"/opt/homebrew/Cellar/go/1.25.6/libexec"} 
INFO [lintersdb] Active 21 linters: [bodyclose errcheck errname errorlint gochecknoinits gofmt goimports goprintffuncname gosec govet ineffassign misspell nilerr nilnil predeclared staticcheck thelper tparallel unparam unused whitespace] 
INFO [loader] Go packages loading at mode 8767 (imports|types_sizes|deps|exports_file|name|compiled_files|files) took 338.086625ms 
INFO [runner/filename_unadjuster] Pre-built 0 adjustments in 16.913458ms 
INFO [linters_context/goanalysis] analyzers took 11.50992502s with top 10 stages: buildir: 5.595256464s, goimports: 1.175858083s, gosec: 492.863875ms, printf: 347.156249ms, ctrlflow: 230.676536ms, errorlint: 207.328082ms, gofmt: 184.538625ms, fact_deprecated: 178.422588ms, buildssa: 141.824709ms, nilness: 127.495631ms 
INFO [runner/exclusion_paths] Skipped 0 issues by pattern "third_party$" 
INFO [runner/exclusion_paths] Skipped 0 issues by pattern "builtin$" 
INFO [runner/exclusion_paths] Skipped 0 issues by pattern "examples$" 
INFO [runner/exclusion_rules] Skipped 11 issues by rules: [Text: "Use of weak random number generator", Path: "_test\\.go"] 
INFO [runner/exclusion_rules] Skipped 5 issues by rules: [Text: "ST1018", Path: "_test\\.go"] 
INFO [runner/exclusion_rules] Skipped 0 issues by rules: [Path: "third_party$", Linters: "gofmt, goimports"] 
INFO [runner/exclusion_rules] Skipped 0 issues by rules: [Path: "builtin$", Linters: "gofmt, goimports"] 
INFO [runner/exclusion_rules] Skipped 0 issues by rules: [Path: "examples$", Linters: "gofmt, goimports"] 
INFO [runner/max_same_issues] 5/8 issues with text "G602: slice index out of range" were hidden, use --max-same-issues 
INFO [runner] Issues before processing: 60, after processing: 3 
INFO [runner] Processors filtering stat (in/out): exclusion_rules: 50/11, fixer: 8/8, uniq_by_line: 8/8, max_same_issues: 8/3, path_prettifier: 3/3, sort_results: 3/3, cgo: 60/60, exclusion_paths: 60/60, max_per_file_from_linter: 8/8, source_code: 3/3, filename_unadjuster: 60/60, nolint_filter: 11/8, max_from_linter: 3/3, severity-rules: 3/3, invalid_issue: 60/60, generated_file_filter: 60/50, diff: 8/8, path_shortener: 3/3, path_absoluter: 60/60, path_relativity: 60/60 
INFO [runner] processing took 1.917292ms with stages: nolint_filter: 1.235417ms, generated_file_filter: 458.834µs, exclusion_rules: 100.459µs, source_code: 44.541µs, exclusion_paths: 37.875µs, path_relativity: 26.833µs, max_same_issues: 3.334µs, sort_results: 2.374µs, invalid_issue: 1.792µs, cgo: 1.334µs, uniq_by_line: 1.249µs, path_absoluter: 791ns, max_from_linter: 583ns, filename_unadjuster: 458ns, path_shortener: 375ns, max_per_file_from_linter: 334ns, fixer: 250ns, path_prettifier: 167ns, diff: 167ns, severity-rules: 125ns 
INFO [runner] linters took 2.229611333s with stages: goanalysis_metalinter: 2.227643375s 
flatten_json.go:756:12: G602: slice index out of range (gosec)
		unescaped[0] = '/'
		         ^
flatten_json.go:758:12: G602: slice index out of range (gosec)
		unescaped[0] = 8
		         ^
flatten_json.go:760:12: G602: slice index out of range (gosec)
		unescaped[0] = 0xc
		         ^
3 issues:
* gosec: 3
INFO File cache stats: 1 entries of total size 24.5KiB 
INFO Memory: 27 samples, avg is 410.1MB, max is 712.0MB 
INFO Execution took 2.59298725s              
```

</details>

### A minimal reproducible example or link to a public repository

<details>

repo is https://github.com/timbray/quamina, file is flatten_json.go

</details>

### Validation

- [x] Yes, I've included all information above (version, config, etc.).

### Supporter

- [ ] I am a sponsor/backer through [GitHub](https://github.com/sponsors/golangci/) or [OpenCollective](https://opencollective.com/golangci-lint)

**Comment by timbray** [comment]
BTW pretty sure this is new with 2.8, that file hasn't changed in a year or more and I just did brew upgrade recently.

**Comment by ldez** [comment]
This is a bug of `gosec`.

https://github.com/securego/gosec/issues/1406

The fix is not inside a gosec release.

Workaround:

```yml
version: "2"
linters:
  settings:
    gosec:
      excludes:
        - G602
```

<details>
<summary>With your configuration</summary>

```yml
version: "2"
linters:
  enable:
    - bodyclose
    - errname
    - errorlint
    - gochecknoinits
    - goprintffuncname
    - gosec
    - misspell
    - nilerr
    - nilnil
    - predeclared
    - staticcheck
    - thelper
    - tparallel
    - unparam
    - whitespace
  settings:
    gosec:
      excludes:
        - G602

  exclusions:
    presets:
      - comments
      - common-false-positives
      - legacy
      - std-error-handling
    rules:
      - path: _test\.go
        text: Use of weak random number generator
      - path: _test\.go
        text: ST1018

formatters:
  enable:
    - gofmt
    - goimports

```

</details>

**Comment by timbray** [comment]
Oh OK thks.

**Comment by vianamjr** [comment]
In an array this should not be a problem:

```go
	var pcs [1]uintptr
	runtime.Callers(2, pcs[:]) // skip caller
	r := slog.NewRecord(time.Now(), slog.LevelError, message, pcs[0])
```

Generates:

```bash
.../service.go:182:63: G602: slice index out of range (gosec)
        r := slog.NewRecord(time.Now(), slog.LevelError, message, pcs[0])
                                                                     ^
```

**Comment by vianamjr** [comment]
There is a PR in the gosec repo that may solve the array situation:
 
https://github.com/securego/gosec/issues/1495#issuecomment-3888587936
